presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,249 @@
+     */
+    public ZOrderCurveSearcher(ZOrder zOrder)
+    {
+        this.zOrder = zOrder;
+
+        // Create mutable version of encodingBits
+        List<Integer> bitLengths = new ArrayList<>(zOrder.encodingBits);
+        this.dimensions = new int[zOrder.maxBitLength];
+        for (int dimensionIndex = zOrder.maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+            for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                if (bitLengths.get(bitLengthIndex) > 0) {
+                    dimensions[dimensionIndex]++;
+                    bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                }
+            }
+        }
+    }
+
+    /**
+     * Searches for and outputs ranges of long addresses within certain ranges in each dimension.
+     */
+    public List<ZAddressRange> zOrderCondensedSearchCurveLongs(List<Integer> minimumValues, List<Integer> maximumValues)
+            throws Exception
+    {
+        return zOrderCondensedSearchCurve(minimumValues, maximumValues, true);
+    }
+
+    /**
+     * Searches for and outputs the long addresses within certain ranges in each dimension.
+     */
+    public List<Long> zOrderSearchCurveLongs(List<Integer> minimumValues, List<Integer> maximumValues)
+            throws Exception
+    {
+        return zOrderSearchCurve(minimumValues, maximumValues, true);
+    }
+
+    /**
+     * Searches for and outputs ranges of integer addresses within certain ranges in each dimension.
+     */
+    public List<ZAddressRange> zOrderCondensedSearchCurveIntegers(List<Integer> minimumValues, List<Integer> maximumValues)
+            throws Exception
+    {
+        return zOrderCondensedSearchCurve(minimumValues, maximumValues, false);
+    }
+
+    /**
+     * Searches for and outputs the integer addresses within certain ranges in each dimension.
+     */
+    public List<Integer> zOrderSearchCurveIntegers(List<Integer> minimumValues, List<Integer> maximumValues)
+            throws Exception
+    {
+        return zOrderSearchCurve(minimumValues, maximumValues, false);
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @param returnLong determines whether to return long addresses or integer addresses
+     * @return the list of address ranges within the given value ranges
+     */
+    private List<ZAddressRange> zOrderCondensedSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues, boolean returnLong)
```

Comment:
This parameter "returnLong" already tells us that we need to consider two inheritance, since we are working on two different types. Since this class only needs to be initialized few times for a query, inheritance cost could be tolerated.